packages/runner-ct/src/iframe/iframes.tsx

```diff
@@ -22,134 +21,135 @@ export function getSpecUrl ({ namespace, spec }, prefix = '') {
+    new IframeModel({
+      state: props.state,
+      restoreDom: autIframe.restoreDom,
+      highlightEl: autIframe.highlightEl,
+      detachDom: autIframe.detachDom,
+      snapshotControls: (snapshotProps) => (
+        <SnapshotControls
+          eventManager={props.eventManager}
+          snapshotProps={snapshotProps}
+          state={props.state}
+          onToggleHighlights={_toggleSnapshotHighlights}
+          onStateChange={_changeSnapshotState}
         />
-        <ScriptError error={scriptError} />
-        <div className='cover' />
-      </div>
-    )
-  }
-
-  componentDidMount () {
-    const config = this.props.config
-
-    this.autIframe = new AutIframe(config)
+      ),
+    }),
+  )
 
-    this.props.eventManager.on('visit:failed', this.autIframe.showVisitFailure)
-    this.props.eventManager.on('before:screenshot', this.autIframe.beforeScreenshot)
-    this.props.eventManager.on('after:screenshot', this.autIframe.afterScreenshot)
-    this.props.eventManager.on('script:error', this._setScriptError)
+  useEffect(() => {
+    props.eventManager.on('visit:failed', autIframe.showVisitFailure)
+    props.eventManager.on('before:screenshot', autIframe.beforeScreenshot)
+    props.eventManager.on('after:screenshot', autIframe.afterScreenshot)
+    props.eventManager.on('script:error', _setScriptError)
 
     // TODO: need to take headless mode into account
     // may need to not display reporter if more than 200 tests
-    this.props.eventManager.on('restart', () => {
-      this._run(this.props.state.spec, config)
+    props.eventManager.on('restart', () => {
+      _run(props.state.spec, props.config)
     })
 
-    this.props.eventManager.on('print:selector:elements:to:console', this._printSelectorElementsToConsole)
+    props.eventManager.on('print:selector:elements:to:console', _printSelectorElementsToConsole)
+
+    props.eventManager.start(props.config)
+
+    setDisposers([
+      autorun(() => {
+        autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
+      }),
+      autorun(() => {
+        autIframe.toggleSelectorHighlight(selectorPlaygroundModel.isShowingHighlight)
+      }),
+      autorun(() => {
+        if (props.state.spec) {
+          _run(props.state.spec, props.config)
+        }
+      }),
+    ])
+
+    iframeModel.listen()
+
+    return () => {
+      props.eventManager.notifyRunningSpec(null)
+      props.eventManager.stop()
+      disposers.forEach((dispose) => {
+        dispose()
+      })
+    }
+  }, [])
 
-    this._disposers.push(autorun(() => {
-      this.autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
-    }))
+  useEffect(() => {
+    props.state.callbackAfterUpdate?.()
+  }, [props.state.callbackAfterUpdate])
 
-    this._disposers.push(autorun(() => {
-      this.autIframe.toggleSelectorHighlight(selectorPlaygroundModel.isShowingHighlight)
-    }))
+  const _toggleSnapshotHighlights = (snapshotProps) => {
```

Comment:
The type of `snapshotProps` is a mystery. I am not sure if the type exists in the code base. So I just left it for now.

Reply:
Yea, I was running into this as well. You could make a simple incomplete type, as we know it at least has the form:

```typescript
interface SnapshotProps {
  snapshots: Array<{ body: unknown }>
  $el?: unknown // Probably HTMLElement or Jquery.Element
}
```

But I'll leave that up to you to decide.

Reply:
I'll spend some time figuring out the entire type at some point. I kind of feel like these partially typed objects do more harm than good - it's often not clear a type is only partially complete, and the reads incorrectly assumes it is.